pennylane/devices/default_qubit.py

```diff
@@ -142,6 +180,148 @@ def _apply_operation(self, operation):
+        """
+        return self._roll(state, 1, axes[0])
+
+    def _apply_y(self, state, axes, **kwargs):
+        """Applies a PauliY gate by adding a negative sign to the 1 index along the axis specified
+        in ``axes``, rolling one unit along the same axis, and multiplying the result by 1j.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        return 1j * self._apply_x(self._apply_z(state, axes), axes)
+
+    def _apply_z(self, state, axes, **kwargs):
+        """Applies a PauliZ gate by adding a negative sign to the 1 index along the axis specified
+        in ``axes``.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        return self._apply_phase(state, axes, -1)
+
+    def _apply_hadamard(self, state, axes, **kwargs):
+        """Apply the Hadamard gate by combining the results of applying the PauliX and PauliZ gates.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        state_x = self._apply_x(state, axes)
+        state_z = self._apply_z(state, axes)
+        return SQRT2INV * (state_x + state_z)
+
+    def _apply_s(self, state, axes, inverse=False):
+        return self._apply_phase(state, axes, 1j, inverse)
+
+    def _apply_t(self, state, axes, inverse=False):
+        return self._apply_phase(state, axes, TPHASE, inverse)
+
+    def _apply_cnot(self, state, axes, **kwargs):
+        """Applies a CNOT gate by slicing along the first axis specified in ``axes`` and then
+        applying an X transformation along the second axis.
+
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        # If axis[1] is larger than axis[0], then state[sl_1] will have self.num_wires - 1 axes with
+        # the target axes shifted down by one. Otherwise, if axis[1] is less than axis[0] then its
+        # axis number in state[sl_1] remains unchanged. For example: state has axes [0, 1, 2, 3] and
+        # axis[0] = 1 and axis[1] = 3. Then, state[sl_1] has axes [0, 1, 2] so that the target axis
+        # has shifted from 3 to 2. If axis[0] = 2 and axis[1] = 1, then state[sl_1] has axes
+        # [0, 1, 2] but with the target axis remaining unchanged.
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
```

Comment:
This part is not entirely clear to me when going for it for some time.

> then state[sl_1] will have self.num_wires - 1 axes with
        # the target axes shifted down by one

Why is it so?

One-piece that is missing is how the first wire is being the control (although the operation does seem to be equivalent). Might be worth explaining perhaps in the main docstring, how using slices and the axes we can end up with the same operation.

Reply:
:thinking: I've tried to explain a bit better, and also add to the main docstring.
7059c5096147436bd933caa1d4410b7956736b32